In [1]:
import pandas as pd
import numpy as np

In [2]:
df = pd.read_csv('https://raw.githubusercontent.com/alexeygrigorev/datasets/master/housing.csv')

In [3]:
df.head(10)

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
0    -122.23     37.88                41.0        880.0           129.0   
1    -122.22     37.86                21.0       7099.0          1106.0   
2    -122.24     37.85                52.0       1467.0           190.0   
3    -122.25     37.85                52.0       1274.0           235.0   
4    -122.25     37.85                52.0       1627.0           280.0   
5    -122.25     37.85                52.0        919.0           213.0   
6    -122.25     37.84                52.0       2535.0           489.0   
7    -122.25     37.84                52.0       3104.0           687.0   
8    -122.26     37.84                42.0       2555.0           665.0   
9    -122.25     37.84                52.0       3549.0           707.0   

   population  households  median_income  median_house_value ocean_proximity  
0       322.0       126.0         8.3252            452600.0        NEAR BAY  
1      2401.0      1138.0         8.3014            358500.0        NEAR BAY  
2       496.0       177.0         7.2574            352100.0        NEAR BAY  
3       558.0       219.0         5.6431            341300.0        NEAR BAY  
4       565.0       259.0         3.8462            342200.0        NEAR BAY  
5       413.0       193.0         4.0368            269700.0        NEAR BAY  
6      1094.0       514.0         3.6591            299200.0        NEAR BAY  
7      1157.0       647.0         3.1200            241400.0        NEAR BAY  
8      1206.0       595.0         2.0804            226700.0        NEAR BAY  
9      1551.0       714.0         3.6912            261100.0        NEAR BAY

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20640 entries, 0 to 20639
Data columns (total 10 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   longitude           20640 non-null  float64
 1   latitude            20640 non-null  float64
 2   housing_median_age  20640 non-null  float64
 3   total_rooms         20640 non-null  float64
 4   total_bedrooms      20433 non-null  float64
 5   population          20640 non-null  float64
 6   households          20640 non-null  float64
 7   median_income       20640 non-null  float64
 8   median_house_value  20640 non-null  float64
 9   ocean_proximity     20640 non-null  object 
dtypes: float64(9), object(1)
memory usage: 1.6+ MB


There are 10 columns in this dataset and 20,640 rows. Of those columns total_bedrooms contains missing Values.

In [5]:
df['ocean_proximity'].value_counts()

ocean_proximity
<1H OCEAN     9136
INLAND        6551
NEAR OCEAN    2658
NEAR BAY      2290
ISLAND           5
Name: count, dtype: int64

Of ocean proximity, there are 5 unique columns.

In [6]:
#use the method im most familiar with
df.loc[:, ['ocean_proximity','median_house_value']].groupby('ocean_proximity').mean()

median_house_value
ocean_proximity                    
<1H OCEAN             240084.285464
INLAND                124805.392001
ISLAND                380440.000000
NEAR BAY              259212.311790
NEAR OCEAN            249433.977427

In [7]:
#use the method taught in the bootcamp; doesnt save as a data frame
df.groupby('ocean_proximity').median_house_value.mean()

ocean_proximity
<1H OCEAN     240084.285464
INLAND        124805.392001
ISLAND        380440.000000
NEAR BAY      259212.311790
NEAR OCEAN    249433.977427
Name: median_house_value, dtype: float64

The average value of the median house value near the bay is 259212

In [8]:
df['total_bedrooms'].isnull().sum()

207

In [9]:
df['total_bedrooms'].mean()

537.8705525375618

In [10]:
df.loc[:, 'total_bedrooms'].fillna(df.loc[:, 'total_bedrooms'].mean(), inplace=True)

In [11]:
df.loc[:, 'total_bedrooms'].isnull().sum()

0

In [12]:
df.loc[:, 'total_bedrooms'].mean()

537.8705525375617

The numbers did not change

In [16]:
island_df = df[df['ocean_proximity'] == 'ISLAND']
island_df = island_df[['housing_median_age', 'total_rooms', 'total_bedrooms']]
island_df

housing_median_age  total_rooms  total_bedrooms
8314                27.0       1675.0           521.0
8315                52.0       2359.0           591.0
8316                52.0       2127.0           512.0
8317                52.0        996.0           264.0
8318                29.0        716.0           214.0

In [17]:
X = island_df.values
XTX = X.T.dot(X)

XTX_inv = np.linalg.inv(XTX)
XTX_inv

array([[ 9.19403586e-04, -3.66412216e-05,  5.43072261e-05],
       [-3.66412216e-05,  8.23303633e-06, -2.77534485e-05],
       [ 5.43072261e-05, -2.77534485e-05,  1.00891325e-04]])

In [19]:
y = np.array([950, 1300, 800, 1000, 1300])
w = (XTX_inv @ X.T) @ y
w[-1]

5.6992294550655656